src/main/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessor.java

```diff
@@ -98,13 +103,36 @@ public void inferenceSentences(
         @NonNull final String modelId,
         @NonNull final List<String> inputText,
         @NonNull final ActionListener<List<List<Float>>> listener
+    ) {
+        inferenceSentencesWithRetry(targetResponseFilters, modelId, inputText, 0, listener);
+    }
+
+    private void inferenceSentencesWithRetry(
+        final List<String> targetResponseFilters,
+        final String modelId,
+        final List<String> inputText,
+        final int retryTime,
+        final ActionListener<List<List<Float>>> listener
     ) {
         MLInput mlInput = createMLInput(targetResponseFilters, inputText);
         mlClient.predict(modelId, mlInput, ActionListener.wrap(mlOutput -> {
             final List<List<Float>> vector = buildVectorFromResponse(mlOutput);
             log.debug("Inference Response for input sentence {} is : {} ", inputText, vector);
             listener.onResponse(vector);
-        }, listener::onFailure));
+        }, e -> {
+            if (shouldRetry(e, retryTime)) {
+                final int retryTimeAdd = retryTime + 1;
+                inferenceSentencesWithRetry(targetResponseFilters, modelId, inputText, retryTimeAdd, listener);
+            } else {
+                listener.onFailure(e);
+            }
+        }));
+    }
+
+    private boolean shouldRetry(Exception e, int retryTime) {
+        final int nodeNotConnectedExceptionIndex = ExceptionUtils.indexOfThrowable(e, NodeNotConnectedException.class);
+        final int nodeDisconnectExceptionIndex = ExceptionUtils.indexOfThrowable(e, NodeDisconnectedException.class);
+        return (nodeDisconnectExceptionIndex != -1 || nodeNotConnectedExceptionIndex != -1) && retryTime < MAX_RETRY;
```

Comment:
I would suggest 1 more simplification here:

```suggestion
    private boolean shouldRetry(Exception e, int retryTime) {
        
        return (RetryUtil.containsException(INFERENCE_API_RETRYABLE_EXCEPTION_LIST, e)) && retryTime < MAX_RETRY;
}

class RetryUtil {
    
    public static boolean containsException(@NonNull final List<Class<?>> exceptionClassesList, final Exception ex) {
     // call below function to check if any exception is present and then return true.
    }
    
    private static boolean containsException(final Exception ex, final ClassClass<?> exceptionClazz) {
         return ExceptionUtils.indexOfThrowable(ex, exceptionClazz) != -1;
    }
}

```

Reply:
In fact, we have analyzed all the exceptions if they're retryable and it doesn't seem to have more exceptions could be retried, otherwise the retry may cause cluster level instability, I wouldn't suggest go too far and over design for this for now.

Reply:
@zane-neo This logic would retry if a model is not loaded, correct?

Reply:
> In fact, we have analyzed all the exceptions if they're retryable and it doesn't seem to have more exceptions could be retried, otherwise the retry may cause cluster level instability, I wouldn't suggest go too far and over design for this for now.

So you are saying there will be no case in future where we need add exception? Thats very strange. I would recommend lets build retry mechanism extensible. 

Also the suggested code is just an example.

Reply:
> @zane-neo This logic would retry if a model is not loaded, correct?

No, this logic only retries the network issue: NodeNotConnectedException and NodeDisconnectedException. When a model is not loaded the requests won't be dispatched to that node at the beginning.

Reply:
> > In fact, we have analyzed all the exceptions if they're retryable and it doesn't seem to have more exceptions could be retried, otherwise the retry may cause cluster level instability, I wouldn't suggest go too far and over design for this for now.
> 
> So you are saying there will be no case in future where we need add exception? Thats very strange. I would recommend lets build retry mechanism extensible.
> 
> Also the suggested code is just an example.

Are you talking about business level retries? These should be retried at underlying layers to avoid network calls. But I can do the refactor to the potential business retries at client side.

Reply:
> Are you talking about business level retries?

No.

I am not saying to refractor any code at client side. I am saying just refractor the code at the Neural Search plugin side as suggested in my older comment.